apps/settings/js/panels/call/call_forwarding.js

```diff
@@ -0,0 +1,202 @@
+/**
+ * Make sure all operations would be executed one by one.
+ *
+ * @module call/call_forwarding
+ */
+define(function(require) {
+  'use strict';
+
+  var Module = require('modules/base/module');
+  var EventEmitter = require('modules/base/event_emitter');
+  var CallUtils = require('panels/call/call_utils');
+  var CallConstant = require('panels/call/call_constant');
+
+  /**
+   * @class CallForwarding
+   * @returns {CallForwarding}
+   */
+  var CallForwarding = Module.create(function() {
+    this.super(EventEmitter).call(this, [
+      'set-callForwarding-invalid',
+      'set-callForwarding-updating',
+      'set-callForwarding-success',
+      'set-callForwarding-error',
+      'update-callForwarding-subpanels',
+      'enable-menuItems',
+      'update-all-desc-info-with-l10n-id',
+      'update-all-desc-info-with-rules'
+    ]);
+
+    this._conn = null;
+
+    /** Stores current states (enabler or not) of the call forwaring reason */
+    this._cfReasonStates = [0, 0, 0, 0];
+  }).extend(EventEmitter);
+
+  CallForwarding.prototype.setCallForwardingValues = function(options) {
+    var key = options.key;
```

Comment:
The api of setting call forwarding is a standard asynchronous function call, so the result of this function can be simply returned by a promise without emitting events.

In this case, an async function is not enough because there are some additional information to display such as "updating". Setting call forwarding takes time, so merely emitting an event could run into problems when users navigate to the panel during setting because the event could be missed. Therefore, in addition to emitting events, a property representing the current state is also necessary.